mozfldp/fed_avg_w_dp.py

```diff
@@ -0,0 +1,250 @@
+    data: Data to train on. In the format of: ([user_1_labels, user_2_labels, ...], [user_1_feats, user_2_feats, ...])
+    """
+
+    # Fix the seed for all rng calls
+    np.random.seed(prms.rand_seed)
+    random.seed(prms.rand_seed)
+
+    user_weights, weight_sum = _init_user_weights_and_weight_sum(
+        prms.num_users, prms.weight_mod
+    )
+    theta = None
+    theta_0 = _init_theta_from_moment_accountant(prms.num_features, prms.num_labels)
+    prev_theta = np.array(theta_0, copy=True)
+    user_sel_prob = 1.0 / prms.num_users
+    standard_dev = _calc_standard_dev(
+        prms.noise_scale, prms.sensitivity, user_sel_prob, weight_sum
+    )  # This feels weird being a constant...
+    user_updates_buf = []
+
+    for round_t in range(prms.num_rounds):
+        # Pick a random set of users to sample
+        random_user_idxs_sample = _get_random_selection_of_user_idxs(prms.num_users)
+
+        # Query the selected users
+        user_updates_buf.clear()
+        for user_idx in random_user_idxs_sample:
+            user_round_labels, user_round_feats = _get_data_for_user_for_round(
+                prms, data, user_idx
+            )
+            user_updates_buf.append(
+                user_update_fed_avg(prms, user_round_feats, user_round_labels, theta_0)
+            )
+
+        # Merge (fc)
+        merged_user_values = _merge_all_user_weights(
+            user_sel_prob, weight_sum, user_updates_buf, user_weights
+        )  # Note that we start at t + 1
+
+        # Note: Assuming for now that S is defined before we run
+        rand_gauss_noise = _gen_gausian_rand_noise(
+            standard_dev, len(merged_user_values)
+        )
+
+        theta = prev_theta + merged_user_values + rand_gauss_noise
+        prev_theta = theta
+
+        print("Theta ([coef, inter] for round {}: \n {}".format(round_t, theta))
+        _moments_accountant_accum_priv_spending(prms.noise_scale)
+
+    privacy_spent = _calc_privacy_spent()
+    print("Total privacy spent from privacy budget: {}".format(privacy_spent))
+
+    coef_slice, inter_slice = _get_coef_and_inter_slice_from_theta(
+        theta, prms.num_features, prms.num_labels
+    )
+
+    return np.array(coef_slice), np.array(inter_slice)
+
+
+# TODO: Give better function name...
+def _merge_all_user_weights(user_sel_prob, weight_sum, user_updates_buf, user_weights):
+    """
+    Merge all user updates for a round into a single delta (vector).
+    """
+
+    num_users_in_batch = len(user_updates_buf)
+    merged_weights = np.zeros(len(user_updates_buf[0]), dtype=np.float64, order="C")
+
+    for i in range(num_users_in_batch):
+        weighted_user_val = np.multiply(user_updates_buf[i], user_weights[i])
+        merged_weights = np.add(merged_weights, weighted_user_val)
+
+    merged_weights = np.divide(merged_weights, user_sel_prob * weight_sum)
+    return merged_weights
+
+
+def flat_clip(sensitivity, vec):
+    vec *= min(1, sensitivity / np.linalg.norm(vec))
```

Comment:
It would be better to use a conditional form, to avoid dividing if the norm is small. Ie:
```python
norm = np.linalg.norm(vec)
if norm > sensitivity:
  vec *= sensitivity / norm
```

Reply:
Optimized!